roc.go

```diff
@@ -0,0 +1,136 @@
+// cutoffs spanning the relative (or receiver)
+// operator characteristic (ROC) curve obtained when y is
+// treated as a binary classifier for classes with weights.
+// Cutoffs are equally spaced from eps less than the minimum value of y
+// to the maximum value of y, including both endpoints -- this means the
+// resulting ROC curve will always begin at (0,0) and end at (1,1).
+// input y must be sorted, and SortWeightedLabelled can be used
+// in order to sort data y together with classes and weights.
+// For a given cutoff value, observations corresponding to
+// entries in y greater than the cutoff are classified as
+// false, and these assigned class labels are compared to
+// classes to calculate FPR and TPR.
+// weights being nil results in the unweighted case.
+// n of zero results in all possible cutoffs being calculated.
+// More details about ROC curves are available at
+// https://en.wikipedia.org/wiki/Receiver_operating_characteristic
+func ROC(n int, y []float64, classes []bool, weights []float64) (tpr, fpr []float64) {
+	if len(y) != len(classes) {
+		panic("stat: slice length mismatch")
+	}
+	if !sort.IsSorted(sort.Float64Slice(y)) {
+		panic("stat: input must be sorted")
+	}
+
+	var iw, tol float64
+	if n != 0 {
+		if n < 2 {
+			panic("stat: cannot less than 2 points on a ROC curve")
+		}
+		iw = (y[len(y)-1] - y[0]) / float64(n-1)
+		if iw <= 0 {
+			panic("stat: y constant or not sorted")
+		}
+		tol = y[0] + iw
+	}
+
+	if n > (len(y) + 1) {
+		tpr = make([]float64, n)
+		fpr = make([]float64, n)
+	} else {
+		tpr = make([]float64, len(y)+1)
+		fpr = make([]float64, len(y)+1)
+	}
+	newEntry := true
+	var j int
+	var nPositive, nNegative float64
+	for i, u := range classes {
+		if weights == nil {
+			// unweighted case
+			if u {
+				nPositive++
+				if newEntry {
+					newEntry = false
+					tpr[j+1] = tpr[j] + 1
+					fpr[j+1] = fpr[j]
+				} else {
+					tpr[j+1]++
+				}
+			} else {
+				if newEntry {
+					newEntry = false
+					tpr[j+1] = tpr[j]
+					fpr[j+1] = fpr[j] + 1
+				} else {
+					fpr[j+1]++
+				}
+			}
+		} else {
+			// weighted case
+			if u {
+				nPositive += weights[i]
+				if newEntry {
+					newEntry = false
+					tpr[j+1] = tpr[j] + weights[i]
+					fpr[j+1] = fpr[j]
+				} else {
+					tpr[j+1] += weights[i]
+				}
+			} else {
+				nNegative += weights[i]
+				if newEntry {
+					newEntry = false
+					tpr[j+1] = tpr[j]
+					fpr[j+1] = fpr[j] + weights[i]
+				} else {
+					fpr[j+1] += weights[i]
+				}
+			}
+		}
+		if i != (len(y) - 1) {
+			if n == 0 {
+				if y[i] != y[i+1] {
+					j++
+					newEntry = true
+				}
+			} else {
+				if y[i+1] > tol {
+					newEntry = true
+					tol += iw
+					j++
+					for y[i+1] > tol {
+						tpr[j+1] = tpr[j]
+						fpr[j+1] = fpr[j]
+						tol += iw
+						j++
+					}
+				}
+			}
+		}
+
+	}
+	if weights == nil {
+		nNegative = 1 / (float64(len(y)) - nPositive)
+	} else {
+		nNegative = 1 / nNegative
+	}
+	nPositive = 1 / nPositive
```

Comment:
This is my approach. Division is more costly that multiplication - in this case is probably doesn't make a great deal of difference, but if I'm dividing by the same value multiple times, I usually multiply by the reciprocal.